### PR TITLE
rgw/pubsub: add amqp push endpoint

### PR DESCRIPTION
**Pull request by yuvalif**
# General Description
This PR is adding support for AMQP0-9-1 ednpoint for push notifications.
- This is based on the existing pubsub framwork (with dedicated RGW on a pubsub zone)
# API
- To define an amqp endpoint the following API should be used:
```
PUT /subscriptions/sub1?topic=tp1&push-endpoint=amqp://guest:guest@localhost:5672/vhost1&amqp-exchange=ex1&amqp-ack-level=broker
```
- This will add a subscription called "sub1", for a previously defined topic "tp1". To an AMQP endpoint (default is 0.9.1, in the future we will add 1.).
- The server run on localhost over the default port (5672), and default rabbitmq user and password (guest/guest) are used.
- We try to connect to a virtual host called "vhost1" (not there by default, the default vhost  is just "/").
- We use only exchanges that route messages based on topics, therefore we validate the existence of an exchange after connection (hence the parameter "amqp-exchange").
# AMQP Acking
- 3 ack levels exist: none (message is considered "delivered" if sent top broker); broker (message is considered "delivered" if acked by broker); routable (message is considered "delivered" if broker can route it - not supported yet))
- in the future we may want to support end2end ack
- currently acking is not reflected back into the sync mechanism or the pubsub acking mechanism. also no retry mechanism exists
- no other behavior was changed in pubsub
# Perf Counters
- Following perf and error counters were added:
```
pubsub_event_triggered: 
Running counter of events with at lease one pubsub topic associated with them
￼
pubsub_event_lost: 
Running counter of events that had  pubsub topics and subscriptions associated with them but that were not stored or pushed to any of the subscriptions

￼pubsub_store_ok: 
Running counter, for all subscriptions, of stored pubsub events 

pubsub_store_fail: 
Running counter, for all subscriptions, of pubsub events that needed to be stored but failed

￼pubsub_events: 
Gauge value of pubsub events in store (not acked or expired) [not implemented yet]

￼pubsub_push_ok: 
Running counter, for all subscriptions, of pubsub events successfully pushed to an endpoint

￼pubsub_push_fail: 
Running counter, for all subscriptions, of pubsub events failed to be pushed to an endpoint

pubsub_push_pending: 
Gauge value of pubsub events pushed to a endpoined bit not acked or nacked yet [not implemented yet]
```
To get these stats the admin socket of the 2nd gateway is needed. Assuming it is under ```out/radosgw.8001.asok```, the command to get all stats would be:
```
ceph --admin-daemon out/radosgw.8001.asok perf dump
```
> Note that ```pubsub_event_triggered``` and ```ipubsub_event_lost``` are incremented per event, while: ```pubsub_store_ok```, ```pubsub_store_fail```, ```pubsub_push_ok```, ```pubsub_push_fail```, are incremented per store/push action on each subscriptions.
# Remaining Work
There are missing features (as noted above and in the TODOs in the code) but also quality Issues that will be addressed:
- logs and some missing stats
- pubsub unit-test
- amqp stubs library (for more lightweight linking when possible)
- integration tests
- performance: caching of the endpoint objects so they are not created for every push (not that the internal connections are already cached)

### CODE REVIEW

**Comment by yuvalif** [comment]
comments forwarded from closed PR #25788:
- investigate: use of exceptions in the code
- investigate: use of fmt lib instead of stringstream for in the "to_string()" functions
- insert amqp namespace to rgw namespace
- std::unordered_map vs. std::vector for holding list of amqp endpoints

File: src/rgw/rgw_sync_module_pubsub.cc [hunk 1]
```diff
@@ -753,32 +747,25 @@ class PSSubscription {
...
                                      pse(_event),
-                                     conf(sub->env->conf),
-                                     sub_conf(sub->sub_conf) {
-      oid_prefix = sub->sub_conf->data_oid_prefix;
+                                     oid_prefix(sub->sub_conf->data_oid_prefix) {
     }
 
     int operate() override {
+      // TODO: in case of "push-only" subscription no need to store event
+      rgw_object_simple_put_params put_obj;
```
Comment (yehudasa) [inline src/rgw/rgw_sync_module_pubsub.cc]: why was this moved from being a class member. Object will be destroyed after first yield. Should be a class member.

File: src/rgw/rgw_pubsub_push.cc [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  if (schema == "http") {
+    return Ptr(new RGWPubSubHTTPEndpoint(endpoint, args));
+  } else if (schema == "https") {
+    // not supported yet
+    return nullptr;
+  } else if (schema == "amqp") {
+    bool exists;
+    std::string version = args.get("amqp-version", &exists);
+    if (!exists) {
+      version = "0-9-1";
```
Comment (yehudasa) [inline src/rgw/rgw_pubsub_push.cc]: these magic versions should really be defined as constants elsewhere

File: src/rgw/rgw_pubsub_push.cc [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  const auto pos = endpoint.find(':');
+  if (pos == std::string::npos) {
+    // unknowm schema
+    return nullptr;
+  }
+  const auto& schema = endpoint.substr(0,pos);
+  if (schema == "http") {
+    return Ptr(new RGWPubSubHTTPEndpoint(endpoint, args));
+  } else if (schema == "https") {
+    // not supported yet
```
Comment (yehudasa) [inline src/rgw/rgw_pubsub_push.cc]: No?

File: src/rgw/rgw_pubsub_push.cc [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+  //fetch the schema from the endpoint
+  const auto pos = endpoint.find(':');
+  if (pos == std::string::npos) {
+    throw configuration_error("malformed endpoint " + endpoint);
+    return nullptr;
+  }
+  const auto& schema = endpoint.substr(0,pos);
+  if (schema == "http") {
+    return Ptr(new RGWPubSubHTTPEndpoint(endpoint, args));
+  } else if (schema == "https") {
```
[the remark this answers is not in the document]
Comment (yehudasa) [reply src/rgw/rgw_pubsub_push.cc]: Why don't we support `https`?

File: src/rgw/rgw_sync_module_pubsub.cc [hunk 5]
```diff
@@ -115,10 +140,19 @@ struct PSSubConfig { /* subscription config */
             const string& default_oid_prefix) {
     name = config["name"];
     topic = config["topic"];
-    push_endpoint = config["push_endpoint"];
+    push_endpoint_name = config["push_endpoint"];
     string default_bucket_name = data_bucket_prefix + name;
     data_bucket_name = config["data_bucket"](default_bucket_name.c_str());
     data_oid_prefix = config["data_oid_prefix"](default_oid_prefix.c_str());
+    if (push_endpoint_name != "") {
```
Comment (yehudasa) [inline src/rgw/rgw_sync_module_pubsub.cc]: `!push_endpoint_name.empty()`

File: src/rgw/rgw_sync_module_pubsub.cc [hunk 6]
```diff
@@ -1142,68 +1161,97 @@ class RGWPSFindBucketTopicsCR : public RGWCoroutine {
...
             ldout(sync_env->cct, 20) << "storing event for subscription=" << *siter << " owner=" << *oiter << " ret=" << retcode << dendl;
             yield call(PSSubscription::store_event_cr(sync_env, sub, event));
             if (retcode < 0) {
               ldout(sync_env->cct, 10) << "ERROR: failed to store event for subscription=" << *siter << " ret=" << retcode << dendl;
-              continue;
+            } else {
+              event_handled = true;
+            }
+            ldout(sync_env->cct, 20) << "push event for subscription=" << *siter << " owner=" << *oiter << " ret=" << retcode << dendl;
+            yield call(PSSubscription::push_event_cr(sync_env, sub, event));
```
Comment (yehudasa) [inline src/rgw/rgw_sync_module_pubsub.cc]: maybe shouldn't call this unconditionally? we know at this point whether we're going to push anything, right?

File: src/rgw/rgw_pubsub_push.cc [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+
+  // PostCR implements async execution of RGWPostHTTPData via coroutine
+  class PostCR : public RGWPostHTTPData, public RGWSimpleCoroutine {
+  private:
+    RGWDataSyncEnv* const sync_env;
+    bufferlist read_bl;
+    const ack_level_t ack_level;
+
+  public:
+    PostCR(const std::string& _post_data,
```
Comment (yehudasa) [inline src/rgw/rgw_pubsub_push.cc]: I really don't like the idea of creating new CRs that do almost exactly the same as the ones in `rgw_cr_rest.h`. We should reuse common infrastructure. This also implements stuff differently, so I'm not sure it's 100% correct.

File: install-deps.sh [hunk 8]
```diff
@@ -238,6 +239,7 @@ else
         echo "Using apt-get to install dependencies"
         $SUDO apt-get install -y devscripts equivs
         $SUDO apt-get install -y dpkg-dev
+        $SUDO apt-get install -y librabbitmq-dev
```
[the remark this answers is not in the document]
Comment (yuvalif) [reply install-deps.sh]: this is in ceph.spec.in

File: src/rgw/rgw_amqp.cc [hunk 9]
```diff
@@ -0,0 +1,240 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+
+//#include "rgw_amqp.h"
```
[the remark this answers is not in the document]
Comment (yuvalif) [reply src/rgw/rgw_amqp.cc]: this is outdated version of the file

File: src/rgw/rgw_amqp.cc [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+#include <cstring>
+#include <map>
+#include <string>
+#include <vector>
+
+// comparisson operator needed so that amqp_connection_info could
+// be used as a key in std::map
+bool operator<(const amqp_connection_info& lhs, const amqp_connection_info& rhs) {
+  // check hostname
+  if (std::strcmp(lhs.host, rhs.host) < 0) return true;
```
[the remark this answers is not in the document]
Comment (yuvalif) [reply src/rgw/rgw_amqp.cc]: this is outdated version of the file. changed the data structure to unordered_map, so no comparison is needed

File: src/rgw/rgw_pubsub_push.cc (hunk 3 above)
Comment (yuvalif) [reply src/rgw/rgw_pubsub_push.cc]: hmm... will give it a try :-)

File: src/rgw/rgw_pubsub_push.cc (hunk 7 above)
Comment (yuvalif) [reply src/rgw/rgw_pubsub_push.cc]: This will require refactoring rgw_cr_rest, which, in turn requires writing unit-tests for it.
Overall, I agree that this is a good idea, but would prefer doing that in a parallel  PR.
Current "PushCR" implementation (which ignores the content of the replies), is 40 lines of code, and based on same infrastructure: RGWPostHTTPData and RGWSimpleCoroutine.
To add different ack levels, I would wait for the refactoring, so code for reading replies is not implemented in multiple places.

File: src/rgw/rgw_amqp.h [hunk 11]
```diff
@@ -5,26 +5,34 @@
 
 #include <string>
 #include <functional>
-#include <stdexcept>
+#include <boost/smart_ptr/intrusive_ptr.hpp>
 
 namespace rgw::amqp {
-// farward declaration of coonection object
+// farward declaration of conection object
```
Comment (mattbenjamin) [inline src/rgw/rgw_amqp.h]: forward :)

File: src/rgw/rgw_amqp.h [hunk 12]
```diff
@@ -14,26 +14,48 @@ struct connection_t;
...
     const std::string& message,
     reply_callback_t cb);
 
 // convert the integer status returned from the "publish" function to a string
 std::string status_to_string(int s);
 
-// return the number of open connections
-unsigned get_connection_number();
+// number of connections
+size_t get_connection_number();
```
Comment (mattbenjamin) [inline src/rgw/rgw_amqp.h]: "connection_number" sounds like an identifier for a unique connection, maybe "connection_count?"

File: src/rgw/rgw_amqp.cc [hunk 13]
```diff
@@ -13,7 +13,7 @@
 #include <vector>
 #include <thread>
 
-namespace amqp {
+namespace rgw::amqp {
```
Comment (mattbenjamin) [inline src/rgw/rgw_amqp.cc]: good idea

File: src/rgw/rgw_amqp.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_amqp.cc]: for this reason, and similar elsewhere, I think it would be good to collapse the series to remove initial development history?

File: src/rgw/rgw_pubsub_push.cc (hunk 7 above)
Comment (mattbenjamin) [reply src/rgw/rgw_pubsub_push.cc]: this seems like fair reasoning?

File: src/rgw/rgw_amqp.h (hunk 11 above)
Comment (mattbenjamin) [reply src/rgw/rgw_amqp.h]: oh, and connection

File: src/rgw/rgw_amqp.cc (hunk 13 above)
Comment (yuvalif) [reply src/rgw/rgw_amqp.cc]: @mattbenjamin  actually - was your comment :-)

**Comment by mattbenjamin** [comment]
@yuvalif can you rebase this?  (sorry!)

**Comment by mattbenjamin** [comment]
@yuval on f28, building from src, found that cmake did not check for compatible rabbitmq library (missing amqp.h)

**Comment by yuvalif** [comment]
@mattbenjamin, library installation for ubuntu was added in install-deps.sh (building and unit testing was tested on ubuntu xenial), so it should not fail in the CI.
will add checks in cmake to indicate that install-deps.sh needs to be called.

**Comment by mattbenjamin** [comment]
Here are the two ci failures, obviously from before your last push:

https://shaman.ceph.com/builds/ceph/wip-rgw-yuval-1/28d640c2c31b27916cae126bae30babbcae345d2/default/142269/
https://shaman.ceph.com/builds/ceph/wip-rgw-yuval-1/28d640c2c31b27916cae126bae30babbcae345d2/default/142268/

**Comment by yuvalif** [comment]
builds in bionic and xenial are no passing:
- https://shaman.ceph.com/builds/ceph/wip-yuval-add_amqp_push_endpoint_new/b2c82c8bd51ac97afcf75aeb6d303b960fcb87be/default/142496/
- https://shaman.ceph.com/builds/ceph/wip-yuval-add_amqp_push_endpoint_new/b2c82c8bd51ac97afcf75aeb6d303b960fcb87be/default/142496/

**Comment by mattbenjamin** [comment]
http://pulpito.ceph.com/yuvalif-2019-02-13_19:49:08-rgw-wip-yuval-add_amqp_push_endpoint_new-distro-basic-smithi/
(I think it's going to be a while;  hint:  submit teuthology jobs with a priority :)

**Comment by yuvalif** [comment]
These are results if running rgw:verify on the branch:
http://pulpito.front.sepia.ceph.com/yuvalif-2019-02-14_16:51:20-rgw:verify-wip-yuval-add_amqp_push_endpoint_new-distro-basic-smithi/
it has 48 fails and 48 passes.
Almost all failures are valgrind issue - finding memleaks in the OSD.
Not sure if these leaks are known or not. Note that code changes in this PR should not affect the OSD.
